[GFC][Integration] Begin allowing grid items with automatic inline sizes.

sammygill opened the pull request:
#### fb308a41a3c173416d5c05adade179cd7909c10a
<pre>
[GFC][Integration] Begin allowing grid items with automatic inline sizes.
<a href="https://bugs.webkit.org/show_bug.cgi?id=303592">https://bugs.webkit.org/show_bug.cgi?id=303592</a>
<a href="https://rdar.apple.com/165883075">rdar://165883075</a>

Reviewed by Vitor Roriz.

In 303843@main we added some initial support for computing the automatic
inline size of grid items. Specifically, support was added for the case
where the grid item is non-replaced, does not have an aspect-ratio, and
its alignment stretches it into the grid area because of normal
alignment.

So in this patch we:
1.) Move the justify alignment checks above the width calculation.
2.) Modify the width check to look at the restrictions above, including
the alignment, in the case where width is auto.

One small change we had to make is that if we are resolving justify-self
against the parent, because it was auto, then we return
ItemPosition::Normal if the grid&apos;s justify-items is legacy to match the
behavior described in:
<a href="https://drafts.csswg.org/css-align-3/#justify-items-property">https://drafts.csswg.org/css-align-3/#justify-items-property</a>

We also need to make the same mapping change in GridFormattingContext so
that the used value of justify-self for the grid item is normal.

Canonical link: <a href="https://commits.webkit.org/306990@main">https://commits.webkit.org/306990@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/acf12939821d74c5a7c3b03a62d743041485b1da

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/143000 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/15473 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/6034 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/151674 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/96221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/f81ae116-03e8-43a5-8e3b-a45feb96f6b7) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/144867 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/16129 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/15554 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/109967 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/96221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/70fc9063-02ad-486a-9587-33dc34c86cc7) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/145949 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/12419 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/127921 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/90877 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/be489a36-e172-4fb4-95ec-c10415739f6b) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/11905 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/9590 "Passed tests") | [✅ 🛠 wpe-libwebrtc](https://ews-build.webkit.org/#/builders/172/builds/1673 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/121292 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/4461 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/153987 "Built successfully") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/15098 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/5106 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/117983 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/15135 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/13082 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/118325 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/30261 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/14277 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/125287 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/70805 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/15141 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/4176 "Passed tests") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/14876 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/78852 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/15084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/14938 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

yulunwu2 commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
```diff
@@ -289,8 +290,36 @@ static OptionSet<GridAvoidanceReason> gridLayoutAvoidanceReason(const RenderGrid
 
-        if (!gridItemStyle->width().isFixed())
-            ADD_REASON_AND_RETURN_IF_NEEDED(GridItemHasNonFixedWidth, reasons, reasonCollectionMode);
+        auto usedJustifySelf = [&]  {
```
nit: I think our gridLayoutAvoidanceReason is getting much harder to follow

sammygill replied:
I did a little bit of cleanup here and moved some of the logic to some helper functions. Hopefully that makes this piece of code easier to read WDYT?

yulunwu2 replied:
LGTM, thank you!

yulunwu2 commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
```diff
@@ -289,8 +290,36 @@ static OptionSet<GridAvoidanceReason> gridLayoutAvoidanceReason(const RenderGrid
+                return GridAvoidanceReason::GridItemHasUnsupportedWidthValue;
+            }
+        );
+        if (gridItemUnsupportedWidthReason == GridAvoidanceReason::GridItemHasUnsupportedWidthValue)
```
do we really need this? It feels like it would be easier to have

            [](const auto&) -> std::optional<GridAvoidanceReason> {
                return GridAvoidanceReason::GridItemHasUnsupportedWidthValue;
            }

become

            [](const auto&) -> std::optional<GridAvoidanceReason> {
            ADD_REASON_AND_RETURN_IF_NEEDED(GridItemHasUnsupportedWidthValue, reasons, reasonCollectionMode);
            }

sammygill replied:
I *think* that would end up returning inside the lambda for the switchOn but we would not end up returning inside `gridLayoutAvoidanceReason` if we needed to?...

yulunwu2 reviewed (commented):
Minor nits around gridLayoutAvoidanceReason

vitorroriz commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
```diff
@@ -163,6 +165,26 @@ static bool hasValidRowEnd(const Style::GridPositionExplicit&, const Style::Grid
+static bool gridItemHasValidWidth(const Style::PreferredSize& width)
+{
+    return WTF::switchOn(width,
+        [&](const CSS::Keyword::Auto&) {
```
why do we capture here?

sammygill replied:
oh we definitely don't need to. probably was typing too fast

sammygill committed:
[GFC][Integration] Begin allowing grid items with automatic inline sizes.
https://bugs.webkit.org/show_bug.cgi?id=303592
rdar://165883075

Reviewed by Vitor Roriz.

In 303843@main we added some initial support for computing the automatic
inline size of grid items. Specifically, support was added for the case
where the grid item is non-replaced, does not have an aspect-ratio, and
its alignment stretches it into the grid area because of normal
alignment.

So in this patch we:
1.) Move the justify alignment checks above the width calculation.
2.) Modify the width check to look at the restrictions above, including
the alignment, in the case where width is auto.

One small change we had to make is that if we are resolving justify-self
against the parent, because it was auto, then we return
ItemPosition::Normal if the grid's justify-items is legacy to match the
behavior described in:
https://drafts.csswg.org/css-align-3/#justify-items-property

We also need to make the same mapping change in GridFormattingContext so
that the used value of justify-self for the grid item is normal.

Canonical link: https://commits.webkit.org/306990@main
